File: erpnext/pms/report/pms_report/pms_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
    columns = get_column(filters)
    data = get_data(filters)
    return columns, data
    
def get_data(filters):
    cond = get_conditions(filters)	
    query = """
        select 
                name,
                employee_name,
                employee,
                pms_group,
                designation,
                pms_calendar,
                approver_name,
                department,
                division,
                region,
                unit,
                section               
            """
    if filters.type == "Target Setup Report":
        query += ",date from `tabTarget Set Up` {}".format(cond)

    if filters.type == "Review Report":
        query += ",review_date from `tabReview` {}".format(cond)

    if filters.type == "Performance Evaluation Report":
        query += ", date_of_joining, reason, evaluation_date, form_i_total_rating, form_ii_total_rating, form_i_score, form_ii_score, final_score, final_score_percent, overall_rating from `tabPerformance Evaluation` {}".format(cond)	
        
    if filters.type == "PMS Summary":
        query += ", date_of_joining, posting_date, form_i_total_rating, form_ii_total_rating, form_i_score, form_ii_score, final_score, final_score_percent, overall_rating from `tabPMS Summary` {}".format(cond)
    
    # frappe.throw(str(query))
    data = frappe.db.sql(query)
    return data

def get_column(filters):
    columns = [
                _("Employee Name") + ":Data:120",
                _("Employee ID") + ":Data:120",
                _("PMS Group") + ":Data:120",
                _("Designation") + ":Data:120",			
                _("PMS Calender") + ":Data:120",
                _("Supervisor") + ":Data:120",
                _("Department") + ":Link/Department:120",
                _("Division") + ":Link/Division:120",
                _("Region") + ":Link/Region:120",
                _("Unit") + ":Link/Unit:120",
                _("Section") + ":Link/Section:120",
                
    ]		
    if filters.get("type") == "Target Setup Report":		
        columns.insert(0,_("Reference") + ":Link/Target Set Up:120"),
        columns.append(("Posting Date") + ":Data:120")		

    if filters.get("type") == "Review Report":		
        columns.insert(0,_("Reference") + ":Link/Review:120"),
        columns.append(("Posting Date") + ":Data:120")	
       
    if filters.get("type") == "Performance Evaluation Report":		
        columns.insert(0,_("Reference") + ":Link/Performance Evaluation:120"),
        columns.append(("Date Of Joining") + ":Date:100"),
        columns.append(("Reason") + ":Data:100"),
        columns.append(("Posting Date") + ":Data:100"),
        columns.append(("Form I Total Rating") + ":Float:100"),				
        columns.append(("Form II Total Rating") + ":Float:100"),		
        columns.append(("Form I Score") + ":Float:100"),				
        columns.append(("Form II Score") + ":Float:100"),				
        columns.append(("Final Score") + ":Float:100"),
        columns.append(("Final Score(%)") + ":Float:100"),				
        columns.append(("Overall Rating") + ":Link/Overall Rating:150")
        
    if filters.get("type") == "PMS Summary":		
        columns.insert(0,_("Reference") + ":Link/PMS Summary:120"),
        columns.append(("Date Of Joining") + ":Date:100"),
        columns.append(("Posting Date") + ":Data:100"),
        columns.append(("Form I Total Rating") + ":Float:100"),				
        columns.append(("Form II Total Rating") + ":Float:100"),		
        columns.append(("Form I Score") + ":Float:100"),				
        columns.append(("Form II Score") + ":Float:100"),				
        columns.append(("Final Score") + ":Float:100"),
        columns.append(("Final Score(%)") + ":Float:100"),				
        columns.append(("Overall Rating") + ":Link/Overall Rating:150")
                    
                           
    return columns

def get_conditions(filters):
    cond = ""
    if filters.pms_calendar:
        cond += " where pms_calendar='{}'".format(filters.pms_calendar)
    
    if filters.type == "Target Setup Report":
        if filters.workflow_state == "Draft":
            cond += " and workflow_state = 'Draft'"
        elif filters.workflow_state == "Waiting Approval":
            cond += " and workflow_state = 'Waiting Approval'"
        elif filters.workflow_state == "Approved":
            cond += " and workflow_state = 'Approved'"
        elif filters.workflow_state == "Rejected":
            cond += " and workflow_state = 'Rejected'"

    elif filters.type == "Review Report":
        if filters.workflow_state == "Draft":
            cond += " and rev_workflow_state = 'Draft'"
        elif filters.workflow_state == "Waiting Approval":
            cond += " and rev_workflow_state = 'Waiting Approval'"
        elif filters.workflow_state == "Approved":
            cond += " and rev_workflow_state = 'Approved'"
        elif filters.workflow_state == "Rejected":
            cond += " and rev_workflow_state = 'Rejected'"
    
    elif filters.type == "Performance Evaluation Report":
        if filters.workflow_state == "Draft":
            cond += " and eval_workflow_state = 'Draft'"
        elif filters.workflow_state == "Waiting Approval":
            cond += " and (eval_workflow_state = 'Waiting Approval' or eval_workflow_state = 'Waiting Supervisor Approval' or eval_workflow_state = 'Waiting PERC')"
        elif filters.workflow_state == "Approved":
            cond += " and eval_workflow_state = 'Approved'"
        elif filters.workflow_state == "Rejected":
            cond += " and eval_workflow_state = 'Rejected'"

    elif filters.type == "PMS Summary":
        if filters.docstatus == "Draft":
            cond += " and docstatus = 0"
        elif filters.docstatus == "Submitted":
            cond += " and docstatus = 1"

    if filters.department:
        cond += " and department='{}' ".format(filters.department)
    if filters.branch:
        cond += " and branch='{}'".format(filters.branch)	
    if filters.division:
        cond += " and division = '{}'".format(filters.division)
    if filters.region:
        cond += " and region = '{}'".format(filters.region)
    if filters.unit:
        cond += " and unit = '{}'".format(filters.unit)
    if filters.from_date:
        cond += " and date_of_joining >= '{}'".format(filters.from_date)
    if filters.to_date:
        cond += " and date_of_joining <= '{}'".format(filters.to_date)
    if filters.reason:
        cond += " and reason = '{}'".format(filters.reason)
    if filters.gender:
        cond += " and gender= '{}'".format(filters.gender)

    return cond

    

File: erpnext/pms/report/pms_report/pms_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["PMS Report"] = {
	"filters": [
			{
				"fieldname": "pms_calendar",
				"label": __("Fiscal Year"),
				"fieldtype": "Link",
				"options": "Fiscal Year",				
				"default": frappe.defaults.get_user_default("fiscal_year"),
				"reqd": 1
		    },			
			{
				"fieldname":"workflow_state",
				"label": __("Status"),
				"fieldtype": "Select",
				"options": ["", "Draft", "Waiting Approval", "Approved", "Rejected"],
				"default":"Approved"
			},
			{
				"fieldname":"docstatus",
				"label": __("Status"),
				"fieldtype": "Select",
				"options": ["", "Draft", "Submitted"],
				"default":"Approved",
				"hidden" : 1
			},	
			{
				"fieldname":"type",
				"label": __("Report Type"),
				"fieldtype": "Select",
				"options": ["Target Setup Report", "Review Report","Performance Evaluation Report","PMS Summary"],
				"default":"Performance Evaluation Report",
				"reqd": 1,
				on_change:function(query_report){
					var type = query_report.get_filter_value('type')
					if (type == "Performance Evaluation Report"){
						var workflow_state = query_report.get_filter("workflow_state"); workflow_state.toggle(true);
						var docstatus = query_report.get_filter("docstatus"); docstatus.toggle(false);
						query_report.get_filter('reason').toggle(type == "Performance Evaluation Report" ? 1:0)
						query_report.get_filter('from_date').toggle(type == "Performance Evaluation Report" ? 1:0)
						query_report.get_filter('to_date').toggle(type == "Performance Evaluation Report" ? 1:0)
					} 
					else if (type == "PMS Summary"){
						var workflow_state = query_report.get_filter("workflow_state"); workflow_state.toggle(false);
						var docstatus = query_report.get_filter("docstatus"); docstatus.toggle(true);
					} 
					else {
						query_report.get_filter('reason').toggle(type == "Performance Evaluation Report" ? 0:1)
						query_report.get_filter('from_date').toggle(type == "Performance Evaluation Report" ? 0:1)
						query_report.get_filter('to_date').toggle(type == "Performance Evaluation Report" ? 0:1)
						var workflow_state = query_report.get_filter("workflow_state"); workflow_state.toggle(true);
						var docstatus = query_report.get_filter("docstatus"); docstatus.toggle(false);
					}
					frappe.query_report.refresh()	
				}
			},
			{
				"fieldname":"branch",
				"label": __("Branch"),
				"fieldtype": "Link",
				"options": "Branch",				
				"reqd": 0
			},
			{
				"fieldname":"department",
				"label": __("Department"),
				"fieldtype": "Link",
				"options": "Department",				
				"reqd": 0
			},
			{
				"fieldname":"division",
				"label": __("Division"),
				"fieldtype": "Link",
				"options": "Division",				
				"reqd": 0
			},
			{
				"fieldname":"region",
				"label": __("Region"),
				"fieldtype": "Link",
				"options": "Region",				
				"reqd": 0
			},	
			{
				"fieldname":"section",
				"label": __("Section"),
				"fieldtype": "Link",
				"options": "Section",				
				"reqd": 0
			},	
			{
				"fieldname":"unit",
				"label": __("Unit"),
				"fieldtype": "Link",
				"options": "Unit",				
				"reqd": 0
			},
			{
				"fieldname":"reason",
				"label": __("Reason"),
				"fieldtype": "Select",
				"options": "\nChange In Section/Division/Department\nSuperannuation/Left\nTransfer\nChange In PMS Group",				
				"hidden": 0
			},
			{
				"fieldname":"from_date",
				"label": __("Appointed From Date"),
				"fieldtype": "Date",
				"hidden":0
			},
			{
				"fieldname":"to_date",
				"label": __("Appointed To Date"),
				"fieldtype": "Date",
				"hidden":0
			},
			{
				"fieldname":"gender",
				"label": __("Gender"),
				"fieldtype": "Select",
				"options": ["","Male", "Female"],				
				"reqd": 0
			},
	]
};
